[ziliangstanley] iP

ziliangstanley opened the pull request:
# Lucian
> Lucian, a Sentinel of Light, is a grim hunter of undying spirits, pursuing them relentlessly and annihilating them with his twin relic pistols. ([source](https://leagueoflegends.fandom.com/wiki/Lucian))

Lucian frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/nus-cs2103-AY2425S2/ip/pull/291).
2. double-click it.
3. add your tasks.
4. let it manage yours tasks for you 👍 

And it is **FREE**!

Features:
- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

ziliangstanley committed:
Change Duke.java to Lucian.java and add greetings dialogue

ziliangstanley committed:
Add echo functionality

ziliangstanley committed:
Add list functionality

ziliangstanley committed:
Create Task class

ziliangstanley committed:
Add mark as done functionality

ziliangstanley committed:
Create the classes of multiple types of tasks

ziliangstanley committed:
Add support for tracking the three types of tasks

ziliangstanley committed:
Add automated text UI testing

ziliangstanley committed:
Create LucianException class

ziliangstanley committed:
Add basic exception handling

ziliangstanley committed:
Add support for deleting tasks from the list

ziliangstanley committed:
Add save function to files

ziliangstanley committed:
Merge branch 'branch-level-7'

ziliangstanley committed:
Add LocalDateTime to files

ziliangstanley committed:
Merge branch 'branch-Level-8'

ziliangstanley committed:
Refactor code to integrate more OOP

ziliangstanley committed:
Merge branch 'A-MOREOOP'

ziliangstanley committed:
Divide classes into packages

ziliangstanley committed:
Merge tag 'A-Packages'

ziliangstanley committed:
Revert "Merge tag 'A-Packages'"

This reverts commit 5685e5b81cbc363ba8d153c8249cfa6284c0c9d0, reversing
changes made to 4d8382e39149b5fa65e02b96f338015fd6032ba7.

ziliangstanley committed:
Reapply "Merge tag 'A-Packages'"

This reverts commit 770205d9345a1ea0cb66f9a9329a3444a8344a1f.

ziliangstanley committed:
Merge branch 'add-gradle-support'

ziliangstanley committed:
Update build.gradle

ziliangstanley committed:
Add JUnit tests

ziliangstanley committed:
Merge branch 'A-JUnit'

ziliangstanley committed:
Package app as a JAR file

ziliangstanley committed:
Add JavaDoc comments

ziliangstanley committed:
Update code to follow coding standard

ziliangstanley committed:
Add search function

ziliangstanley committed:
Merge branch 'branch-A-JavaDoc'

ziliangstanley committed:
Merge branch 'branch-A-CodingStandard'

ziliangstanley committed:
Merge branch 'branch-Level-9'

ziliangstanley committed:
Implement Checkstyle files

ziliangstanley committed:
Fix all checkstyle errors less javadoc comments

ziliangstanley committed:
Merge branch 'branch-A-CheckStyle'

ziliangstanley committed:
Add GUI to application

ziliangstanley committed:
Merge branch 'branch-Level-10'

nicholasohjj commented on src/main/java/lucian/Lucian.java:
```diff
@@ -0,0 +1,156 @@
+ */
+public class Lucian {
+    private static final String FILE_NAME = "./data/tasks.txt";
+    private Ui ui;
+    private Storage storage;
+    private TaskList tasks;
```
consider making the fields `final `if they are not intended to be reassigned after initialization

nicholasohjj commented on src/main/java/lucian/Lucian.java:
```diff
@@ -0,0 +1,156 @@
+        try {
+            tasks = new TaskList(storage.load());
+        } catch (IOException e) {
+            tasks = new TaskList();
```
consider adding a log statement or a more descriptive error message when catching `IOException`

nicholasohjj commented on src/main/java/lucian/Lucian.java:
```diff
@@ -0,0 +1,156 @@
+            case "list":
+                return ui.showMessage(tasks.printTasks());
+            case "mark":
+                int markIndex = Integer.parseInt(words[1]) - 1;
```
Consider adding validation for `words[1]` in commands like `mark`, `unmark`, and `delete `to ensure the input is a valid integer

nicholasohjj commented on src/main/java/lucian/Lucian.java:
```diff
@@ -0,0 +1,156 @@
+            default:
+                return ui.showMessage("You did not give me a valid command...");
+            }
+        } catch (Exception e) {
```
The catch (`Exception e`) block is too broad. Consider catching specific exceptions

nicholasohjj commented on src/main/java/lucian/Lucian.java:
```diff
@@ -0,0 +1,156 @@
+     * @return The created {@code Task} object.
+     * @throws LucianException If the task format is invalid.
+     */
+    private Task createTask(String input) throws LucianException {
```
Avoid hardcoding values like 5, 9, 6, etc. Consider defining constants for these offsets to improve readability and maintainability.

nicholasohjj commented on src/main/java/lucian/Lucian.java:
```diff
@@ -0,0 +1,156 @@
+
+            try {
+                createdTask = new Event(description, java.time.LocalDate.parse(fromString),
+                        java.time.LocalDate.parse(toString));
```
The date parsing logic is repeated for `Deadline `and `Event`. Consider extracting this into a helper method to reduce redundancy.

nicholasohjj commented on src/main/java/lucian/Lucian.java:
```diff
@@ -0,0 +1,156 @@
+                createdTask = new Event(description, java.time.LocalDate.parse(fromString),
+                        java.time.LocalDate.parse(toString));
+            } catch (java.time.format.DateTimeParseException e) {
+                throw new LucianException("Use YYYY-MM-DD format...");
```
consider making them more user-friendly. For example, instead of "Use YYYY-MM-DD format...", you could say "Please enter the date in YYYY-MM-DD format."

nicholasohjj reviewed (changes requested):
I like how the code is modular, with clear separation of concerns between `Ui`, `Storage`, and `TaskList`.

ziliangstanley committed:
Add assertions to ensure valid command inputs and task creation

Assertions have been added to check for valid indices in mark, unmark,
and delete commands. Additionally, assertions now ensure that
task descriptions are non-empty before task creation.

- Ensure commands requiring parameters have valid inputs.
- Prevent creation of ToDo, Deadline, and Event tasks with empty descriptions.
- Validate date parsing for Deadline and Event tasks.

Assertions are enabled by running the program with `-ea` flag.
They serve as internal checks to detect logical errors early.

ziliangstanley committed:
Refactor handleCommand() to improve readability and maintainability

The handleCommand() method was deeply nested and contained repetitive
logic, making it difficult to read and maintain. Additionally,
hardcoded magic strings were used for task types, reducing code clarity.

- Extracted command handling logic into separate helper methods.
- Reduced deep nesting in handleCommand() by using early returns.
- Replaced magic strings with constants for better maintainability.

These changes enhance readability, make the code easier to modify,
and ensure better adherence to coding best practices.

ziliangstanley committed:
Merge pull request #1 from ziliangstanley/branch-A-Assertions

Add assertions to ensure valid command inputs and task creation

ziliangstanley committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/lucian/Lucian.java

ziliangstanley committed:
Merge pull request #2 from ziliangstanley/branch-A-CodeQuality

Branch a code quality

ziliangstanley committed:
Add stats command to display task statistics

ziliangstanley committed:
Merge branch 'branch-BCD-Extension'

ziliangstanley committed:
Add more exception handling for features

ziliangstanley committed:
Add representative screenshot

ziliangstanley committed:
Update README.md

ziliangstanley committed:
Update build.gradle

ziliangstanley committed:
Revert "Update build.gradle"

This reverts commit f32e87fc8ce6b012f7aa05c443191d1686560621.

ziliangstanley committed:
Merge remote-tracking branch 'origin/master'

ziliangstanley committed:
Update build.gradle

ziliangstanley committed:
Refactor code with more OOP

ziliangstanley committed:
Merge pull request #5 from ziliangstanley/refactor-OOP

Refactor code with more OOP
